src/Elastic.Apm.AspNetCore/ApmMiddleware.cs

```diff
@@ -39,14 +40,24 @@ public async Task InvokeAsync(HttpContext context)
 				Raw = context.Request?.Path.Value //TODO
 			};
 
-			transaction.Context.Request = new Request( context.Request.Method, url)
+			Dictionary<string, object> requestHeaders = null;
+			if (Agent.Config.CaptureHeaders)

```

Comment:
We should pass an `IConfigurationReader` instance to `ApmMiddleware` here and not use the static `Agent.Config` property. With that we can inject the reader during testing. The problem with using the static instance is that we have tests running in parallel. On my machine when I run all the tests then `HomeSimplePageTransactionTest` fails on this assert:
```
if (_agent.ConfigurationReader.CaptureHeaders)
{
    transaction.Context.Response.Headers.Should().NotBeNull(); //<- this remains null
```

If I only run `HomeSimplePageTransactionTest` then it's green. (btw. I'm thinking about maybe getting rid of the `.ConfigurationReader` property... )

(e.g. `ReadConfingsFromEnvVarsViaIConfig` sets environment variables and the static instance defaults to reading those, if this happens in parallel, that can be problem, but once we dependency inject a new instance this problem is resolved).

So we can just add an `IConfigurationReader` constructor parameter to `ApmMiddleware` and use that one. [Here](https://github.com/elastic/apm-agent-dotnet/blob/master/src/Elastic.Apm.AspNetCore/ApmMiddlewareExtension.cs#L40) we already have one, we can just add a parameter to [this method](https://github.com/elastic/apm-agent-dotnet/blob/master/src/Elastic.Apm.AspNetCore/ApmMiddlewareExtension.cs#L51) where we can pass it to the `ApmMiddleware`. Additionally the previous method is used in tests, so each test can inject a new instance.

Reply:
You are right. I didn't want to do that because I tought it can be major. I'm changing it like that ;-) and try to put more test on it.